### ISSUE TITLE
CA1508 false positive in pattern matching

### ISSUE THREAD

**Issue by yaakov-h** [dotnet/roslyn-analyzers#4590]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 5.0.1 (Latest)

### Steps To Reproduce

Compile the following C# code snippet with CA1508 enabled.

```csharp
using System;
using System.IO;
using System.Reflection;

namespace CA1508Repro
{
    public abstract class DeploymentDefinition
    {
    }

    public class DeploymentConfigurationException : Exception
    {
        public DeploymentConfigurationException(string message) : base(message) { }
        public DeploymentConfigurationException(string message, Exception innerException) : base(message, innerException) { }
    }

    public class Test
    {
        string binariesPath = "";

        DeploymentDefinition LoadDefinition(string assemblyName, string typeName)
        {
            object? instance;

            try
            {
                var assemblyToLoad = Path.Combine(binariesPath, assemblyName);
                var assembly = Assembly.LoadFrom(assemblyToLoad);
                var type = assembly.GetType(typeName, throwOnError: true)!;
                instance = Activator.CreateInstance(type);
            }
            catch (Exception ex)
            {
                throw new DeploymentConfigurationException(FormattableString.Invariant($"Failed to create instance of '{typeName}' from assembly '{assemblyName}'."), ex);
            }

            if (!(instance is DeploymentDefinition definition))
            {
                throw new DeploymentConfigurationException(FormattableString.Invariant($"'{typeName}' must be a subclass of DeploymentDefinition in order to be used as a custom deployment definiton."));
            }

            return definition;
        }
    }
}
```

### Expected behavior

No diagnostics from CA1508.

### Actual behavior

```
Class1.cs(37,19,37,62): warning CA1508: 'instance is DeploymentDefinition definition' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
```

**Comment by mavasani** [comment]
Verified that this has been fixed in master and does not repro on https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers/5.0.3
